### ISSUE TITLE
Getting "Parse Error" when using PHP hooks

### ISSUE THREAD

**Issue by jackbentley** [apiaryio/dredd#699]
#### Describe your problem

Random error on execution. Using `php` hooks. After exit, the hooks service is left running and can't be re-run because the port is in use.

Seems persistent for some reason.

```
error: Error: Parse Error
    at Error (native)
    at Socket.socketOnData (_http_client.js:363:20)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:548:20)
{ ... lots of tests running fine ... }
warn: Hook handling timed out.
/{...}/node_modules/drafter.js/lib/drafter.nomem.js:1
(function (exports, require, module, __filename, __dirname) { ((function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.drafter=f()}}))((function(){var module;var Module={"ready":false,"log":(function(text){console.log("drafter.js: "+text)}),"logErr":(function(text){console.error("drafter.js: "+text)}),"onRuntimeInitialized":(function(){this.ready=true})};var Module;if(!Module)Module=(typeof Module!=="undefined"?Module:null)||{};var moduleOverrides={};for(var key in Module){if(Module.hasOwnProperty(key)){moduleOverrides[key]=Module[key]}}var ENVIRONMENT_IS_WEB=false;var ENVIRONMENT_IS_WORKER=false;var ENVIRONMENT_IS_NODE=false;var ENVIRONMENT_IS_SHELL=false;if(Module["ENVIRONMENT"]){if(Module["EN

Error: Callback was already called.
    at /{...}/node_modules/async/dist/async.js:837:32
    at /{...}/node_modules/async/dist/async.js:1010:13
    at TransactionRunner.emitResult (/{...}/node_modules/dredd/lib/transaction-runner.js:444:14)
    at /{...}/node_modules/dredd/lib/transaction-runner.js:120:34
    at TransactionRunner.runHooksForData (/{...}/node_modules/dredd/lib/transaction-runner.js:191:14)
    at /{...}/node_modules/dredd/lib/transaction-runner.js:115:32
    at /{...}/node_modules/dredd/lib/transaction-runner.js:188:16
    at /{...}/node_modules/async/dist/async.js:1013:9
    at /{...}/node_modules/async/dist/async.js:356:16
    at replenish (/{...}/node_modules/async/dist/async.js:877:25)
    at iterateeCallback (/{...}/node_modules/async/dist/async.js:867:17)
    at /{...}/node_modules/async/dist/async.js:840:16
    at /v/node_modules/async/dist/async.js:1010:13
    at /{...}/node_modules/dredd/lib/transaction-runner.js:168:24
    at /{...}/node_modules/dredd/lib/transaction-runner.js:283:18
    at EventEmitter.messageHandler (/{...}/node_modules/dredd/lib/hooks-worker-client.js:333:20)
    at emitOne (events.js:96:13)
    at EventEmitter.emit (events.js:188:7)
    at Socket.<anonymous> (/{...}/node_modules/dredd/lib/hooks-worker-client.js:286:44)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:548:20)
```

#### What command line options do you use?

```bash
$ dredd
```

#### What is in your `dredd.yml`?

```yaml
dry-run: null
hookfiles: tests/hooks/*.php
language: vendor/bin/dredd-hooks-php
sandbox: false
server: null
server-wait: 3
init: false
custom: {}
names: false
only: []
reporter: []
output: []
header: []
sorted: false
user: null
inline-errors: false
details: false
method: []
color: true
level: info
timestamp: false
silent: false
path: []
hooks-worker-timeout: 5000
hooks-worker-connect-timeout: 1500
hooks-worker-connect-retry: 500
hooks-worker-after-connect-wait: 100
hooks-worker-term-timeout: 5000
hooks-worker-term-retry: 500
hooks-worker-handler-host: localhost
hooks-worker-handler-port: 61321
config: ./dredd.yml
blueprint: tests/api-description.apib
endpoint: 'http://test.dev/api/'
```

#### What's your `dredd --version` output?

```
dredd v2.2.5 (Linux 3.10.0-514.2.2.el7.x86_64; x64)
```

#### Does `dredd --level=debug` uncover something?

Not really...

```
debug: Running hooks...
debug: Dredd received some data from hooks handler.
debug: Emitting to reporters: test start
debug: Requesting tested server errored: Error: Parse Error
    at Error (native)
    at Socket.socketOnData (_http_client.js:363:20)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:548:20)
debug: Emitting to reporters: test error
error: DELETE /some/url duration: 9ms
error: Error: Parse Error
    at Error (native)
    at Socket.socketOnData (_http_client.js:363:20)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:548:20)
{ ... lots of tests running fine ... }
/{...}/node_modules/drafter.js/lib/drafter.nomem.js:1
(function (exports, require, module, __filename, __dirname) { ((function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.drafter=f()}}))((function(){var module;var Module={"ready":false,"log":(function(text){console.log("drafter.js: "+text)}),"logErr":(function(text){console.error("drafter.js: "+text)}),"onRuntimeInitialized":(function(){this.ready=true})};var Module;if(!Module)Module=(typeof Module!=="undefined"?Module:null)||{};var moduleOverrides={};for(var key in Module){if(Module.hasOwnProperty(key)){moduleOverrides[key]=Module[key]}}var ENVIRONMENT_IS_WEB=false;var ENVIRONMENT_IS_WORKER=false;var ENVIRONMENT_IS_NODE=false;var ENVIRONMENT_IS_SHELL=false;if(Module["ENVIRONMENT"]){if(Module["EN

Error: Callback was already called.
    at /{...}/node_modules/async/dist/async.js:837:32
    at /{...}/node_modules/async/dist/async.js:1010:13
    at TransactionRunner.emitResult (/{...}/node_modules/dredd/lib/transaction-runner.js:464:12)
    at /{...}/node_modules/dredd/lib/transaction-runner.js:120:34
    at TransactionRunner.runHooksForData (/{...}/node_modules/dredd/lib/transaction-runner.js:191:14)
    at /{...}/node_modules/dredd/lib/transaction-runner.js:115:32
    at /{...}/node_modules/dredd/lib/transaction-runner.js:188:16
    at /{...}/node_modules/async/dist/async.js:1013:9
    at /{...}/node_modules/async/dist/async.js:356:16
    at replenish (/{...}/node_modules/async/dist/async.js:877:25)
    at iterateeCallback (/{...}/node_modules/async/dist/async.js:867:17)
    at /{...}/node_modules/async/dist/async.js:840:16
    at /{...}/node_modules/async/dist/async.js:1010:13
    at /{...}/node_modules/dredd/lib/transaction-runner.js:168:24
    at /{...}/node_modules/dredd/lib/transaction-runner.js:283:18
    at EventEmitter.messageHandler (/{...}/node_modules/dredd/lib/hooks-worker-client.js:333:20)
    at emitOne (events.js:96:13)
    at EventEmitter.emit (events.js:188:7)
    at Socket.<anonymous> (/{...}/node_modules/dredd/lib/hooks-worker-client.js:286:44)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:548:20)
```

#### Can you send us failing test in a Pull Request?
No - don't know whats causing the error.

#### Extra

My hooks file contains a database population on a `beforeAll` (which seems to be working fine) and a simple `beforeEach` which sets a header to a string.

**Comment by honzajavorek** [comment]
@jackbentley 

> Not really...

It actually does! Thanks! According to the `dredd --level=debug` output it looks like the problem is in the communication between Dredd and the PHP hooks (cc @ddelnano). It seems like Dredd gets something from the PHP hooks in the socket message, which cannot be parsed as valid JSON.

**Comment by ddelnano** [comment]
@honzajavorek hm I have not seen this before.  Also its not like the php hooks have had any changes to it anytime recently.

@jackbentley what version of the php hooks are you using?  Can you share you hookfiles?

**Comment by jackbentley** [comment]
@honzajavorek It looked like the same errors shown without `debug` to my un-trained-dredd eye 😄 . Is there anywhere I can make it output this to confirm?

@ddelnano Looks like it might be related to my `beforeAll` as uncommenting it seems to make the errors go away (except when it appeared once? so maybe not? or it might be some sort of caching that caused that?).

Sometimes it appears twice, and some times appears once. And usually doesn't force-exit (from the `Callback was already called`), but does sometimes.

The `beforeAll` builds my container, then loads fixtures via [nelmio/alice](https://github.com/nelmio/alice) using the `NativeLoader`.

My `beforeEach` is just this at the moment:
```
Hooks::beforeEach(function (&$transaction) {
  $transaction->request->headers->Cookie = "PHPSESSID=sessionId";
});
```

Maybe theres something var_dumping/echoing somewhere (can't find anything though)? Or because it's taking a little more time to execute?

**Comment by ddelnano** [comment]
Can you share your beforeAll?

**Comment by jackbentley** [comment]
Here you go. Had to remove some specifics, but essentially all the same:

```php
use App\Assistant;
use App\Database;
use App\EntityManager;
use App\Kernel;
use Dredd\Hooks;

require_once __DIR__ . "/../../autoload.php";

/** @var Nelmio\Alice\Loader\NativeLoader|null $loader */
$loader = null;
/** @var string|null $fixtures */
$fixtures = null;
/** @var EntityManager|null $entityManager */
$entityManager = null;

function setup()
{
  global $loader, $fixtures, $entityManager;

  $app = new Kernel('test', 'test', false);
  $container = $app->buildContainer();
  $entityManager = $container->get('entity.manager');
  $database = $container->get('database');
  $loader = new Nelmio\Alice\Loader\NativeLoader();
  $fixtures = __DIR__ . '/../fixtures';

  $assistant = new Assistant();
  $database->exec("DROP TABLE IF EXISTS table1, table2, table3");
  $assistant->writeDatabaseStructure($database);
}

function loadFixture($file)
{
  global $loader, $fixtures, $entityManager;

  $results = $loader->loadFile($fixtures . '/' . $file);

  foreach ($results->getObjects() as $name => $object) {
    $repo = $entityManager->getRepositoryByObject($object);

    if (!$repo) {
      throw new \RuntimeException(sprintf("Couldn't find repository for '%s'", gettype($object)));
    }

    $repo->create($object);
  }
}

Hooks::beforeAll(function (&$transaction) {
  setup();

  loadFixture('fixtures1.yml');
  loadFixture('fixtures2.yml');
  loadFixture('fixtures3.yml');
});
```

**Comment by jackbentley** [comment]
Ok, looking at the error again, I noticed it's saying `Requesting tested server errored`. The test that the error keeps occuring on is expected to return a `204`. It looks like dredd is failing because of the no content? It seems to be failing regardless of if the response code is `204` or not.

The blueprint for the action is just this simple:
```markdown
### Delete User [DELETE]

+ Response 204
```

**Comment by jackbentley** [comment]
Fixed the tests now.

Looks like my server was putting something into the response body on a `204`. Probably something like `0`. Forced it to output nothing on a `204` and no more parse errors.

The tests were still failing, because my server was returning a `Content-Type` header on a `204` response. Removing that made everything pass.

So, looks like dredd is unable to pull the response body on a 204, to show you it failed?

**Comment by honzajavorek** [comment]
@jackbentley I made the assumption from the output you posted and knowledge of [Dredd's execution life cycle](http://dredd.readthedocs.io/en/latest/how-it-works/#execution-life-cycle).

You are right the problem might be with 204 - I didn't realise that from the bug report at first. There are issues related to empty content and 204: https://github.com/apiaryio/dredd/issues/468 / https://github.com/apiaryio/dredd/issues/196 / https://github.com/apiaryio/dredd/issues/556

Could you please confirm the problem you have is already described in those?

**Comment by honzajavorek** [comment]
I believe this is already addressed by [v4.7.1](https://github.com/apiaryio/dredd/releases/tag/v4.7.1) and further, I just forgot to close this one. Thanks everyone involved and I'm sorry it took so long to fix this.

It would be awesome if you could verify the issues you had are no longer relevant. Thanks!
